File: python/day01/day01.py
def part1():
    with open('../../haskell/data/day01-input.txt') as f:

        def parse_int(line: str):
            digits = [c for c in line if c.isdigit()]
            return int(digits[0] + digits[-1])

        return sum([parse_int(line) for line in f])


digit_by_name = {"one": 1, "two": 2, "three": 3, "four": 4,
                 "five": 5, "six": 6, "seven": 7, "eight": 8, "nine": 9}


def match_digit_name(s: str):
    for name in digit_by_name:
        if s[:len(name)] == name:
            return digit_by_name[name]
    return -1


def part2():
    with open('../../haskell/data/day01-input.txt') as f:
        def parse_int(line: str):
            def try_parse(substr: str) -> int:
                if substr[0].isdigit():
                    return int(substr[0])
                return match_digit_name(substr)

            digits = [try_parse(line[idx:]) for idx, _ in enumerate(line)]
            digits = [d for d in digits if d >= 0]
            return int(digits[0]*10 + digits[-1])
        return sum([parse_int(line) for line in f])


if __name__ == '__main__':
    print(f"part 1: {part1()}")
    print(f"part 2: {part2()}")

File: haskell/src/Day01.hs
module Day01 where

import Data.Char
import Data.List
import Data.Maybe
import Data.Maybe (catMaybes)
import Paths_aoc2023 (getDataFileName)
import Text.Printf (printf)

day01 :: IO ()
day01 = do
  inputLines <- lines <$> (getDataFileName "day01-input.txt" >>= readFile)
  printf "part 1: %d\n" (part1 inputLines)
  printf "part 2: %d\n" (part2 inputLines)

part1 :: [String] -> Int
part1 lines =
  let digits = filter isDigit
      firstAndLast line = read [head (digits line), last (digits line)]
   in sum $ map firstAndLast lines :: Int

nameToDigit :: String -> Maybe Int
nameToDigit name
  | "one" `isPrefixOf` name = Just 1
  | "two" `isPrefixOf` name = Just 2
  | "three" `isPrefixOf` name = Just 3
  | "four" `isPrefixOf` name = Just 4
  | "five" `isPrefixOf` name = Just 5
  | "six" `isPrefixOf` name = Just 6
  | "seven" `isPrefixOf` name = Just 7
  | "eight" `isPrefixOf` name = Just 8
  | "nine" `isPrefixOf` name = Just 9
  | otherwise = Nothing

enumerate = zip [0 ..]

part2 :: [String] -> Int
part2 lines =
  let parseSubstr s =
        if isDigit $ head s
          then Just $ digitToInt $ head s
          else nameToDigit s
      digits line = catMaybes [parseSubstr (drop n line) | (n, _) <- enumerate line]
      firstAndLast line = 10 * head (digits line) + last (digits line)
   in sum $ map firstAndLast lines :: Int
